### ISSUE TITLE
Never ending request

### ISSUE THREAD

**Issue by rolios** [Netflix/pollyjs#248]
## Description

I use PollyJS with Puppeteer to record and replay http requests in E2E test context.
I have one never-ending http call, I'm trying to figure out why. I didn't managed to find the root cause of it and started to wonder if it could be a bug.

- Every other requests works as expected, only this one doesn't.
- the request is an http POST to a laravel backend to get an auth token: `/oauth/token`
- Based on body parameters, if username and password are invalid, the backend answer with an error and PollyJS behave correctly
- If username and password are valid, the backend answer code 200 with a token
- Request works as expected when Polly is not enabled and request interception is off

### Manual tests in devtools 

```js
fetch(url, {
        method: 'POST',
        headers: {
	    accept: 'application/json',
            'content-type': 'application/json;charset=UTF-8',
        },
        body: JSON.stringify(credentials),
    }
```

#### Without polly

<img width="972" alt="Capture d'écran 2019-08-06 18 47 32" src="https://user-images.githubusercontent.com/1358738/62550095-b09b4f00-b87a-11e9-9e15-529c922dfd2d.png"><img width="765" alt="Capture d'écran 2019-08-06 18 48 23" src="https://user-images.githubusercontent.com/1358738/62550138-cad52d00-b87a-11e9-9ce6-d0ca58ec3168.png">

#### With polly

<img width="1011" alt="Capture d'écran 2019-08-06 18 41 52" src="https://user-images.githubusercontent.com/1358738/62549645-edb31180-b879-11e9-8cb9-e4188fb984de.png">
Details of the passthrough request:
<img width="846" alt="Capture d'écran 2019-08-06 18 51 31" src="https://user-images.githubusercontent.com/1358738/62550384-3c14e000-b87b-11e9-9963-99c0b7e719ba.png"><img width="649" alt="Capture d'écran 2019-08-06 19 10 29" src="https://user-images.githubusercontent.com/1358738/62551913-df66f480-b87d-11e9-89d9-c61acc5d57ad.png">

I tried to understand what is happening by watching page events in the polly-puppeteer-adapter and have seen my request going through these events : `request / response / request failed (timeout)`

### Dependencies

```json
{
    "@pollyjs/adapter-puppeteer": "^2.6.0",
    "@pollyjs/core": "^2.6.0",
    "@pollyjs/persister-fs": "^2.6.0",
}
```
### Config

```js
new PollyJS('', {
      mode: 'passthrough',
      adapters: ['puppeteer'],
      persister: 'fs',
});
```

## Environment

```
Node.js v11.5.0
darwin 18.7
6.10.2
1.17.3
```

**Comment by offirgolan** [comment]
@rolios any chance you can create a working reproduction that I can play around with?

**Comment by rolios** [comment]
@offirgolan ok, I'm preparing that  👍

**Comment by rodrigogs** [comment]
I'm havin the exact same problem:
```js
    const { page } = this.helpers.Puppeteer;
    await page.setRequestInterception(true);

    const polly = new Polly(title, {
      mode: 'record',
      adapters: ['puppeteer'],
      persister: 'fs',
      adapterOptions: {
        puppeteer: { page },
      },
      persisterOptions: {
        fs: {
          recordingsDir: path.resolve(__dirname, '..', '..', 'e2e-output', 'requests'),
        },
      },
    }
```

**Comment by offirgolan** [comment]
@rodrigogs can you create a working reproduction that I can use to debug?

**Comment by rodrigogs** [comment]
@offirgolan turns out that the hanging problem was due to the `persister-fs` module, which is another problem. The following code is not working for me anyway:

```javascript
const { Polly } = require('@pollyjs/core');
const PuppeteerAdapter = require('@pollyjs/adapter-puppeteer');
// const FSPersister = require('@pollyjs/persister-fs');
const puppeteer = require('puppeteer');

Polly.register(PuppeteerAdapter);
// Polly.register(FSPersister);

(async () => {
  try {
    const browser = await puppeteer.launch({ headless: false });
    const page = await browser.newPage();

    await page.setRequestInterception(true);

    new Polly('Test', {
      mode: 'record',
      adapters: ['puppeteer'],
      // persister: ['fs'],
      adapterOptions: {
        puppeteer: { page },
      },
    });

    await page.goto('https://accounts.google.com/ServiceLogin');
    await page.waitFor('input[name=identifier]');
    await page.$eval('input[name=identifier]', el => el.value = 'rodrigo.smscom@gmail.com');
    await page.click('#identifierNext');
  } catch (err) {
    console.error(err);
  }
})();
```

When I'm intercepting requests with Polly, the password page fails to load, when I disable Polly, it works perfectly. If you uncomment the `persister-fs` code, will see that there's also a problem with `adapter-puppeteer` along with `persister-fs`.

**Comment by jasonmit** [comment]
@rodrigogs what version of puppeteer?  I know this used to occur on an older version of puppeteer.

**Comment by rodrigogs** [comment]
@jasonmit version `1.19.0`

**Comment by offirgolan** [comment]
@rodrigogs would you mind checking if this issue occurs for you on `v1.10.0`?

**Comment by rodrigogs** [comment]
https://gist.github.com/rodrigogs/7db680570b04b0d2808cf482b2b5548c

Same behavior. Tested with `1.10.0` and `1.20.0`.

**Comment by jasonmit** [comment]
I'm able to reproduce this and see that a request is being iniatiated but not captured by Polly.  Since `await page.setRequestInterception(true);` is set puppeteer is waiting to be told to continue.  Since Polly is unaware of the request, the intercepted request is never responded to and eventually times out.

![image](https://user-images.githubusercontent.com/3108309/65906861-1a663f80-e378-11e9-83b1-ba5f1eb76829.png)

I don't yet have a solve for this but figured I'd drop some more details.

**Comment by jasonmit** [comment]
Some more findings, the passthrough request that polly makes (the one immediately under the timed out request) responds back with an empty body.  However, the request does have a response body if you make the request via `curl`.  For some reason the same request made via `fetch` returns back nothing from the API other than a 200 status code.

**Comment by jasonmit** [comment]
Yeah I think the approach of going over `fetch` just isn't compatible with all scenarios.  We need to be able to trigger XHR events (i.e., `progress`, `load`, `loadend`, `error`, `abort`) and we lose that with the current implementation of having the request go over fetch.

**Comment by witqq** [comment]
any workarounds?

**Comment by jasonmit** [comment]
Do you have a reproduction that doesn't involve a Google request?  Have not triaged further and would appreciate any additional information.

**Comment by szymn** [comment]
I started to see a similar issue when I added `Cache-Control: no-store` header to the resource I was fetching while using pollyjs through codeceptjs.

Package versions:
```
  "@codeceptjs/mock-request": "0.3.0",
  "codeceptjs": "2.6.8",
  "puppeteer": "3.0.4",
  "@pollyjs/adapter-puppeteer": "4.3.0",
  "@pollyjs/core": "4.3.0",
  "@pollyjs/persister-fs": "4.3.0"
```
